### PR TITLE
Implement add parameter support for target typed new

### LINKED ISSUE #42559: AddParameter and target-typed new
Adding an argument to an implicit object creation doesn't trigger AddParameter.

Similarly, GenerateConstructor doesn't trigger on `C c = new(1, 2, 3, 4);` to offer to add a constructor with 4 parameters.

```
class C
{
    C(int i) { }
    void M()
    {
       C c = new(1 $$); // typing a second argument here should trigger AddParameter
    }
}
```

Relates to https://github.com/dotnet/roslyn/issues/28489 (target-typed new feature)

### PR DESCRIPTION
**Pull request by allisonchou**
Fixes #42559.

(Created a separate issue #47928 for the other bug mentioned in the original issue for clarity.)

### CODE REVIEW

**Review by davidwengier — APPROVED** [review]
It's interesting to me that it needed this much change. Is it possible for `SyntaxFacts.GetObjectCreationType` to do something different, so it could support implicit creation in the same way? Though getting it to return something such that GetSymbolInfo returns an INamedTypeSymbol might be impossible, I'm just thinking out loud :)

**Comment by CyrusNajmabadi** [comment]
> It's interesting to me that it needed this much change. Is it possible for SyntaxFacts.GetObjectCreationType to do something different, so it could support implicit creation in the same way?

SyntaxFacts is just supposed to be purely syntactic, and only make it possible to treat *really* equivalent VB and C# constructs the same.  So, in vb and C# tehre is an ObjectCreationExpressionSyntax.  So SytnaxFacts allows you to operate over those same constructs agnostically.  It is *not* for treating different language constructs in teh same language the same way.

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 1]
```diff
@@ -140,15 +140,30 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
             {
 
                 // Not supported if this is "new { ... }" (as there are no parameters at all.
-                var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
+                var typeNode = syntaxFacts.IsImplicitObjectCreationExpression(node) ?
+                    node : syntaxFacts.GetObjectCreationType(objectCreation);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs:144]: either:

```suggestion
                var typeNode = syntaxFacts.IsImplicitObjectCreationExpression(node)
                    ? node : syntaxFacts.GetObjectCreationType(objectCreation);
```

or 

```suggestion
                var typeNode = syntaxFacts.IsImplicitObjectCreationExpression(node)
                    ? node
                    : syntaxFacts.GetObjectCreationType(objectCreation);
```

File: src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs [hunk 2]
```diff
@@ -140,15 +140,30 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                // Implicit object creation expressions
+                if (symbol is IMethodSymbol methodSymbol)
+                {
+                    type = methodSymbol.ContainingType;
+                }
+                // Regular object creation expressions
+                else if (symbol is INamedTypeSymbol namedTypeSymbol)
+                {
+                    type = namedTypeSymbol;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs:162]:
```suggestion
var type = symbol switch
{
    IMethodSymbol methodSymbol => methodSymbol.ContainingType,
    ...,
    _ => null,
};
```

**Review by CyrusNajmabadi — APPROVED** [review]
LGTM with the suggestions below.

**Pull request merged**